### PR TITLE
PXD-1960 fix(message): fix error message

### PR DESCRIPTION
**Pull request by giangbui**
See following error:

 

Traceback (most recent call last):
{{ File "/sheepdog/sheepdog/transactions/upload/_init_.py", line 38, in single_transaction_worker}}
{{ transaction.parse_doc(*doc_args)}}
{{ File "/sheepdog/sheepdog/transactions/upload/transaction.py", line 81, in parse_doc}}
{{ self.parse_entities(data)}}
{{ File "/sheepdog/sheepdog/transactions/upload/transaction.py", line 112, in parse_entities}}
{{ self.instantiate()}}
{{ File "/sheepdog/sheepdog/transactions/upload/transaction.py", line 163, in instantiate}}
{{ entity.instantiate()}}
{{ File "/sheepdog/sheepdog/transactions/upload/entity.py", line 149, in instantiate}}
{{ self.node = self.get_node_merge()}}
{{ File "/sheepdog/sheepdog/transactions/upload/sub_entities.py", line 180, in get_node_merge}}
{{ node = super(FileUploadEntity, self).get_node_merge()}}
{{ File "/sheepdog/sheepdog/transactions/upload/entity.py", line 323, in get_node_merge}}
{{ return self.get_node_create()}}
{{ File "/sheepdog/sheepdog/transactions/upload/sub_entities.py", line 158, in get_node_create}}
{{ self._populate_files_from_index()}}
{{ File "/sheepdog/sheepdog/transactions/upload/sub_entities.py", line 377, in _populate_files_from_index}}
{{ self.file_by_hash = self.get_file_from_index_by_hash()}}
{{ File "/sheepdog/sheepdog/transactions/upload/sub_entities.py", line 534, in get_file_from_index_by_hash}}
{{ document = self.transaction.signpost.get_with_params(params)}}
{{ File "/sheepdog/src/indexclient/indexclient/client.py", line 118, in get_with_params}}
{{ raise e}}
HTTPError: 400 Client Error: size must be an integer for url: http://indexd-service/index/?limit=1&hash=md5%3A1E%2B31&size=11.7

Comments
Ascending order - Click to sort in descending order
Permalink
zflamig Zachary Flamig added a comment - 26/Sep/18 1:10 PM

When submitting a file size that isn't an integer (say a float like 11.7) this will generate an internal server error. We should output a better error message instead.

This occurred on NIAID with the imaging_file node.

### CODE REVIEW

**Comment by giangbui** [comment]
The error occurs when sheepdog tries to retrieve/create a document with non-integer file size which fails indexd validation.

File: sheepdog/transactions/upload/sub_entities.py [hunk 1]
```diff
@@ -531,7 +532,10 @@ def get_file_from_index_by_hash(self):
             # if `document` exists, `document.did` is the UUID that is already
             # registered in indexd for this entity.
             if params:
-                document = self.transaction.signpost.get_with_params(params)
+                try:
+                    document = self.transaction.signpost.get_with_params(params)
+                except requests.HTTPError as e:
+                    raise UserError(code=e.response.status_code, message=e.message)
```
Comment (philloooo) [inline sheepdog/transactions/upload/sub_entities.py]: we should add some context to this error message:
`Fail to register this data node in indexd: {message}` ?

**Comment by zflamig** [comment]
Can you test this on dev and paste the error message here that you get now @giangbui? We need to make sure it is self explanatory enough that a user can figure out what is wrong + how to fix it. It looks like the JSON validation happens after this step, but since we require a few fields we may need to prevalidate that them before querying indexd...

**Comment by giangbui** [comment]
{
    "message": "Fail to register the data node in indexd. Detail 400 Client Error: size must be an integer for url: http://indexd-service/index/?limit=1&hash=md5%3A584460781d7bc4d287fd2b95668eb3f8&size=4.1"
}

**Comment by zflamig** [comment]
This message is okay, but not great. Should we do some pre-validation @philloooo ?

**Comment by philloooo** [comment]
@giangbui can you check if the json validation in pre_validate() can be moved to before instantiate() ?

**Comment by giangbui** [comment]
Yeah. We can call `json_validator.record_errors(self.entities)` before instantiate the entities. The validation method does nothing but validating json docs; Thus it does not require any info from `instantiate` function as well as the code block in `pre_validate`
Here is the error message looks like:
[
    {
        "action": null,
        "errors": [
            {
                "keys": [
                    "file_size"
                ],
                "message": "3.3 is not of type 'integer'",
                "type": "INVALID_VALUE"
            }
        ],
        "id": null,
        "related_cases": [],
        "type": "mutation_panel",
        "unique_keys": [
            {
                "project_id": "DEV-test",
                "submitter_id": "mutation_panel_001"
            }
        ],
        "valid": false,
        "warnings": []
    }
]

**Pull request merged**